lib/ts/nextjs.ts

```diff
@@ -127,6 +128,145 @@ export default class NextJS {
+        return result;
+    }
+
+    static async withSession<NextRequest extends PartialNextRequest, NextResponse extends Response>(
+        req: NextRequest,
+        handler: (session: SessionContainer | undefined) => Promise<NextResponse>,
+        options?: VerifySessionOptions
+    ) {
+        const query = Object.fromEntries(new URL(req.url).searchParams.entries());
+        const cookies: Record<string, string> = Object.fromEntries(
+            req.cookies.getAll().map((cookie) => [cookie.name, cookie.value])
+        );
+
+        let baseRequest = new PreParsedRequest({
+            method: req.method as HTTPMethod,
+            url: req.url,
+            query: query,
+            headers: req.headers,
+            cookies: cookies,
+            getFormBody: () => req!.formData(),
+            getJSONBody: () => req!.json(),
+        });
+
+        const { session, nextResponse, baseResponse } = await NextJS.commonSSRSession(baseRequest, options);
+
+        if (nextResponse) {
+            return nextResponse as NextResponse;
+        }
+
+        if (!baseResponse) {
+            throw Error("Expected baseResponse to be present");
+        }
+
+        let userResponse = await handler(session);
+
+        let didAddCookies = false;
+        let didAddHeaders = false;
+
+        for (const respCookie of baseResponse.cookies) {
+            didAddCookies = true;
+            userResponse.headers.append(
+                "Set-Cookie",
+                serialize(respCookie.key, respCookie.value, {
+                    domain: respCookie.domain,
+                    expires: new Date(respCookie.expires),
+                    httpOnly: respCookie.httpOnly,
+                    path: respCookie.path,
+                    sameSite: respCookie.sameSite,
+                    secure: respCookie.secure,
+                })
+            );
+        }
+
+        baseResponse.headers.forEach((value: string, key: string) => {
+            didAddHeaders = true;
+            userResponse.headers.set(key, value);
+        });
+
+        /**
+         * For some deployment services (Vercel for example) production builds can return cached results for
+         * APIs with older header values. In this case if the session tokens have changed (because of refreshing
+         * for example) the cached result would still contain the older tokens and sessions would stop working.
+         *
+         * As a result, if we add cookies or headers from base response we also set the Cache-Control header
+         * to make sure that the final result is not a cached version.
+         */
+        if (didAddCookies || didAddHeaders) {
+            if (!userResponse.headers.has("Cache-Control")) {
+                // This is needed for production deployments with Vercel
+                userResponse.headers.set("Cache-Control", "no-cache, no-store, max-age=0, must-revalidate");
+            }
+        }
+
+        return userResponse;
+    }
 }
```

Comment:
got me to approve this PR, i would also need to see changes in the docs that happen cause of this addition. So you can make changes in docs by using this branch of the node SDK, and then we can see.

Reply:
done - https://github.com/supertokens/docs/pull/765